File: js/thread-utils.ts
/**
 * Thread utilities for LLM-as-a-judge scorers.
 *
 * This module provides utilities for working with preprocessed conversation
 * messages (threads) in LLM scorer templates.
 */

/**
 * A message with role and content fields (LLM chat message format).
 */
export interface LLMMessage {
  role: string;
  content: unknown;
}

function isObject(value: unknown): value is { [key: string]: unknown } {
  return value instanceof Object && !(value instanceof Array);
}

/**
 * Check if an item looks like an LLM message (has role and content).
 */
export function isRoleContentMessage(item: unknown): item is LLMMessage {
  return isObject(item) && "role" in item && "content" in item;
}

/**
 * Check if a value is an array of LLM messages.
 */
export function isLLMMessageArray(value: unknown): value is LLMMessage[] {
  return Array.isArray(value) && value.every(isRoleContentMessage);
}

function indent(text: string, prefix = "  "): string {
  return text
    .split("\n")
    .map((line) => (line ? prefix + line : prefix))
    .join("\n");
}

function truncateMiddle(text: string, maxLen: number): string {
  if (text.length <= maxLen) return text;
  const charsRemoved = text.length - maxLen + 30;
  const ellipsis = ` [...${charsRemoved} chars truncated...] `;
  const avail = maxLen - ellipsis.length;
  if (avail <= 0) return text.slice(0, maxLen);
  const left = Math.floor(avail / 2);
  const right = avail - left;
  return text.slice(0, left) + ellipsis + text.slice(-right);
}

interface PendingToolCall {
  name: string;
  args: string;
}

function isTypedPart(
  part: unknown,
): part is { type: string; [key: string]: unknown } {
  return isObject(part) && typeof part.type === "string";
}

function extractToolCalls(content: unknown[]): Map<string, PendingToolCall> {
  const toolCalls = new Map<string, PendingToolCall>();

  for (const part of content) {
    if (!isTypedPart(part) || part.type !== "tool_call") continue;

    const id = typeof part.tool_call_id === "string" ? part.tool_call_id : "";
    if (!id) continue;

    const name =
      typeof part.tool_name === "string" ? part.tool_name : "unknown";

    let args = "";
    if (isObject(part.arguments)) {
      const argsObj = part.arguments;
      if (argsObj.type === "valid") {
        args = JSON.stringify(argsObj.value);
      } else if (typeof argsObj.value === "string") {
        args = argsObj.value;
      } else {
        args = JSON.stringify(argsObj.value);
      }
    }

    toolCalls.set(id, { name, args });
  }

  return toolCalls;
}

function unwrapContent(content: unknown): string {
  if (typeof content === "string") {
    try {
      const parsed = JSON.parse(content);
      return unwrapContent(parsed);
    } catch {
      const errorMatch = content.match(/^error:\s*'(.+)'$/s);
      if (errorMatch) {
        return errorMatch[1];
      }
      return content;
    }
  }

  if (Array.isArray(content)) {
    const textParts: string[] = [];
    for (const item of content) {
      if (isObject(item) && typeof item.text === "string") {
        textParts.push(unwrapContent(item.text));
      } else if (typeof item === "string") {
        textParts.push(unwrapContent(item));
      }
    }
    if (textParts.length > 0) {
      return textParts.join("\n");
    }
  }

  if (isObject(content) && typeof content.text === "string") {
    return unwrapContent(content.text);
  }

  return typeof content === "string" ? content : JSON.stringify(content);
}

function formatToolResult(
  toolCallId: string,
  toolName: string,
  output: unknown,
  pendingToolCalls: Map<string, PendingToolCall>,
): string {
  const pendingCall = pendingToolCalls.get(toolCallId);
  const name = toolName || pendingCall?.name || "tool";
  const args = pendingCall?.args || "";

  const resultContent = unwrapContent(output);
  const lines = [`Tool (${name}):`];

  if (args) {
    lines.push(`  Args:`);
    lines.push(`    ${truncateMiddle(args, 500)}`);
  }

  const isError =
    resultContent.toLowerCase().includes("error:") ||
    resultContent.toLowerCase().includes('"error"') ||
    resultContent.toLowerCase().startsWith("error");

  if (isError) {
    lines.push(`  Error:`);
    lines.push(`    ${truncateMiddle(resultContent, 500)}`);
  } else {
    lines.push(`  Result:`);
    lines.push(`    ${truncateMiddle(resultContent, 500)}`);
  }

  if (pendingCall) {
    pendingToolCalls.delete(toolCallId);
  }

  return lines.join("\n");
}

function formatToolResults(
  content: unknown[],
  pendingToolCalls: Map<string, PendingToolCall>,
): string[] {
  const results: string[] = [];

  for (const part of content) {
    if (!isTypedPart(part) || part.type !== "tool_result") continue;

    const toolCallId =
      typeof part.tool_call_id === "string" ? part.tool_call_id : "";
    const toolName = typeof part.tool_name === "string" ? part.tool_name : "";

    results.push(
      formatToolResult(toolCallId, toolName, part.output, pendingToolCalls),
    );
  }

  return results;
}

function extractTextContent(content: unknown): string {
  if (typeof content === "string") {
    return content.trim() ? content : "";
  }

  if (!Array.isArray(content)) {
    return "";
  }

  const parts: string[] = [];
  for (const part of content) {
    if (typeof part === "string" && part.trim()) {
      parts.push(part);
    } else if (isTypedPart(part)) {
      if (part.type === "text" && typeof part.text === "string") {
        parts.push(part.text);
      } else if (part.type === "reasoning" && typeof part.text === "string") {
        parts.push(`[thinking: ${part.text.slice(0, 100)}...]`);
      }
    } else if (isObject(part) && typeof part.text === "string") {
      parts.push(part.text);
    }
  }

  return parts.join("\n");
}

/**
 * Format an array of LLM messages as human-readable text.
 */
export function formatMessageArrayAsText(messages: LLMMessage[]): string {
  const pendingToolCalls = new Map<string, PendingToolCall>();
  for (const msg of messages) {
    if (msg.role === "assistant" && Array.isArray(msg.content)) {
      const calls = extractToolCalls(msg.content);
      for (const [id, call] of calls) {
        pendingToolCalls.set(id, call);
      }
    }
  }

  const parts: string[] = [];
  for (const msg of messages) {
    const role = msg.role;
    const capitalizedRole = role.charAt(0).toUpperCase() + role.slice(1);

    if (role === "tool" && Array.isArray(msg.content)) {
      const toolResults = formatToolResults(msg.content, pendingToolCalls);
      parts.push(...toolResults);
    } else {
      const text = extractTextContent(msg.content);
      if (text) {
        parts.push(`${capitalizedRole}:\n${indent(text)}`);
      }
    }
  }

  return parts.join("\n\n");
}

/**
 * Template variables computed from a thread for use in LLM-as-a-judge scorers.
 */
export interface ThreadTemplateVars {
  thread: unknown[];
  thread_text: string;
  thread_count: number;
  first_message: unknown | null;
  last_message: unknown | null;
  user_messages: unknown[];
  assistant_messages: unknown[];
  human_ai_pairs: Array<{ human: unknown; assistant: unknown }>;
}

/**
 * Compute template variables from a thread for use in mustache templates.
 * Uses lazy getters so expensive computations only run when accessed.
 */
export function computeThreadTemplateVars(
  thread: unknown[],
): ThreadTemplateVars {
  let _thread_text: string | undefined;
  let _user_messages: unknown[] | undefined;
  let _assistant_messages: unknown[] | undefined;
  let _human_ai_pairs:
    | Array<{ human: unknown; assistant: unknown }>
    | undefined;

  return {
    thread,
    thread_count: thread.length,

    get thread_text(): string {
      if (_thread_text === undefined) {
        if (isLLMMessageArray(thread)) {
          _thread_text = formatMessageArrayAsText(thread);
        } else {
          _thread_text = thread
            .map((item) =>
              typeof item === "string" ? item : JSON.stringify(item),
            )
            .join("\n");
        }
      }
      return _thread_text;
    },

    get first_message(): unknown | null {
      return thread[0] ?? null;
    },

    get last_message(): unknown | null {
      return thread[thread.length - 1] ?? null;
    },

    get user_messages(): unknown[] {
      if (_user_messages === undefined) {
        _user_messages = thread.filter(
          (m) => isRoleContentMessage(m) && m.role === "user",
        );
      }
      return _user_messages;
    },

    get assistant_messages(): unknown[] {
      if (_assistant_messages === undefined) {
        _assistant_messages = thread.filter(
          (m) => isRoleContentMessage(m) && m.role === "assistant",
        );
      }
      return _assistant_messages;
    },

    get human_ai_pairs(): Array<{ human: unknown; assistant: unknown }> {
      if (_human_ai_pairs === undefined) {
        _human_ai_pairs = [];
        const users = thread.filter(
          (m) => isRoleContentMessage(m) && m.role === "user",
        );
        const assistants = thread.filter(
          (m) => isRoleContentMessage(m) && m.role === "assistant",
        );
        const pairCount = Math.min(users.length, assistants.length);
        for (let i = 0; i < pairCount; i++) {
          _human_ai_pairs.push({ human: users[i], assistant: assistants[i] });
        }
      }
      return _human_ai_pairs;
    },
  };
}

File: py/autoevals/thread_utils.py
"""Thread utilities for LLM-as-a-judge scorers.

This module provides helpers for deriving template variables from conversation
threads and formatting thread values for Mustache rendering.
"""

from __future__ import annotations

import json
import re
from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any

THREAD_VARIABLE_NAMES = [
    "thread",
    "thread_count",
    "first_message",
    "last_message",
    "user_messages",
    "assistant_messages",
    "human_ai_pairs",
]

# Match variables after "{{" or "{%" (e.g., {{thread}}, {{ thread }}, {% if thread %}).
THREAD_VARIABLE_PATTERN = re.compile(r"\{[\{%]\s*(" + "|".join(THREAD_VARIABLE_NAMES) + r")")


def template_uses_thread_variables(template: str) -> bool:
    return bool(THREAD_VARIABLE_PATTERN.search(template))


def is_role_content_message(item: Any) -> bool:
    return isinstance(item, Mapping) and "role" in item and "content" in item


def is_llm_message_array(value: Any) -> bool:
    return isinstance(value, list) and all(is_role_content_message(item) for item in value)


def filter_system_messages_from_thread(thread: list[Any]) -> list[Any]:
    return [message for message in thread if not (isinstance(message, Mapping) and message.get("role") == "system")]


def _indent(text: str, prefix: str = "  ") -> str:
    return "\n".join(prefix + line for line in text.split("\n"))


def _truncate_middle(text: str, max_len: int) -> str:
    if len(text) <= max_len:
        return text
    chars_removed = len(text) - max_len + 30
    ellipsis = f" [...{chars_removed} chars truncated...] "
    avail = max_len - len(ellipsis)
    if avail <= 0:
        return text[:max_len]
    left = avail // 2
    right = avail - left
    return text[:left] + ellipsis + text[-right:]


def _is_typed_part(part: Any) -> bool:
    return isinstance(part, Mapping) and isinstance(part.get("type"), str)


@dataclass
class _PendingToolCall:
    name: str
    args: str


def _extract_tool_calls(content: list[Any]) -> dict[str, _PendingToolCall]:
    tool_calls: dict[str, _PendingToolCall] = {}
    for part in content:
        if not _is_typed_part(part) or part["type"] != "tool_call":
            continue

        tool_call_id = part.get("tool_call_id")
        if not isinstance(tool_call_id, str) or not tool_call_id:
            continue

        tool_name = part.get("tool_name") if isinstance(part.get("tool_name"), str) else "unknown"

        args = ""
        args_obj = part.get("arguments")
        if isinstance(args_obj, Mapping):
            if args_obj.get("type") == "valid":
                args = json.dumps(args_obj.get("value"))
            else:
                value = args_obj.get("value")
                if isinstance(value, str):
                    args = value
                else:
                    args = json.dumps(value)

        tool_calls[tool_call_id] = _PendingToolCall(name=tool_name, args=args)

    return tool_calls


def _unwrap_content(content: Any) -> str:
    if isinstance(content, str):
        try:
            return _unwrap_content(json.loads(content))
        except Exception:
            error_match = re.match(r"^error:\s*'(.+)'$", content, flags=re.DOTALL)
            if error_match:
                return error_match.group(1)
            return content

    if isinstance(content, list):
        text_parts: list[str] = []
        for item in content:
            if isinstance(item, Mapping) and isinstance(item.get("text"), str):
                text_parts.append(_unwrap_content(item["text"]))
            elif isinstance(item, str):
                text_parts.append(_unwrap_content(item))
        if text_parts:
            return "\n".join(text_parts)

    if isinstance(content, Mapping) and isinstance(content.get("text"), str):
        return _unwrap_content(content["text"])

    return content if isinstance(content, str) else json.dumps(content)


def _format_tool_result(
    tool_call_id: str,
    tool_name: str,
    output: Any,
    pending_tool_calls: dict[str, _PendingToolCall],
) -> str:
    pending_call = pending_tool_calls.get(tool_call_id)
    name = tool_name or (pending_call.name if pending_call else "tool")
    args = pending_call.args if pending_call else ""

    result_content = _unwrap_content(output)
    lines = [f"Tool ({name}):"]

    if args:
        lines.append("  Args:")
        lines.append(f"    {_truncate_middle(args, 500)}")

    is_error = (
        "error:" in result_content.lower()
        or '"error"' in result_content.lower()
        or result_content.lower().startswith("error")
    )
    if is_error:
        lines.append("  Error:")
        lines.append(f"    {_truncate_middle(result_content, 500)}")
    else:
        lines.append("  Result:")
        lines.append(f"    {_truncate_middle(result_content, 500)}")

    if pending_call:
        del pending_tool_calls[tool_call_id]

    return "\n".join(lines)


def _format_tool_results(content: list[Any], pending_tool_calls: dict[str, _PendingToolCall]) -> list[str]:
    results: list[str] = []
    for part in content:
        if not _is_typed_part(part) or part["type"] != "tool_result":
            continue

        tool_call_id = part.get("tool_call_id") if isinstance(part.get("tool_call_id"), str) else ""
        tool_name = part.get("tool_name") if isinstance(part.get("tool_name"), str) else ""
        results.append(_format_tool_result(tool_call_id, tool_name, part.get("output"), pending_tool_calls))

    return results


def _extract_text_content(content: Any) -> str:
    if isinstance(content, str):
        return content if content.strip() else ""
    if not isinstance(content, list):
        return ""

    parts: list[str] = []
    for part in content:
        if isinstance(part, str) and part.strip():
            parts.append(part)
        elif _is_typed_part(part):
            if part["type"] == "text" and isinstance(part.get("text"), str):
                parts.append(part["text"])
            elif part["type"] == "reasoning" and isinstance(part.get("text"), str):
                parts.append(f"[thinking: {part['text'][:100]}...]")
        elif isinstance(part, Mapping) and isinstance(part.get("text"), str):
            parts.append(part["text"])

    return "\n".join(parts)


def format_message_array_as_text(messages: list[Mapping[str, Any]]) -> str:
    pending_tool_calls: dict[str, _PendingToolCall] = {}
    for msg in messages:
        if msg.get("role") == "assistant" and isinstance(msg.get("content"), list):
            for tool_call_id, pending in _extract_tool_calls(msg["content"]).items():
                pending_tool_calls[tool_call_id] = pending

    parts: list[str] = []
    for msg in messages:
        role = msg.get("role", "")
        if not isinstance(role, str):
            role = str(role)
        capitalized_role = role[:1].upper() + role[1:]
        content = msg.get("content")

        if role == "tool" and isinstance(content, list):
            parts.extend(_format_tool_results(content, pending_tool_calls))
        else:
            text = _extract_text_content(content)
            if text:
                parts.append(f"{capitalized_role}:\n{_indent(text)}")

    return "\n\n".join(parts)


class RenderableMessage(dict[str, Any]):
    def __str__(self) -> str:
        role = self.get("role", "")
        content = self.get("content")
        content_str = content if isinstance(content, str) else json.dumps(content)
        return f"{role}: {content_str}"


class RenderableMessageArray(list[Any]):
    def __str__(self) -> str:
        return format_message_array_as_text(self)


def _to_renderable_message(message: Mapping[str, Any]) -> RenderableMessage:
    return RenderableMessage(message)


def _to_renderable_message_array(messages: list[Any]) -> RenderableMessageArray:
    wrapped: list[Any] = []
    for message in messages:
        if is_role_content_message(message):
            wrapped.append(_to_renderable_message(message))
        else:
            wrapped.append(message)
    return RenderableMessageArray(wrapped)


def compute_thread_template_vars(thread: list[Any]) -> dict[str, Any]:
    renderable_thread = _to_renderable_message_array(thread) if is_llm_message_array(thread) else thread

    first_message = renderable_thread[0] if len(renderable_thread) > 0 else None
    last_message = renderable_thread[-1] if len(renderable_thread) > 0 else None

    user_messages = [
        message for message in renderable_thread if is_role_content_message(message) and message.get("role") == "user"
    ]
    assistant_messages = [
        message
        for message in renderable_thread
        if is_role_content_message(message) and message.get("role") == "assistant"
    ]
    pair_count = min(len(user_messages), len(assistant_messages))
    human_ai_pairs = [{"human": user_messages[idx], "assistant": assistant_messages[idx]} for idx in range(pair_count)]

    return {
        "thread": renderable_thread,
        "thread_count": len(thread),
        "first_message": first_message,
        "last_message": last_message,
        "user_messages": _to_renderable_message_array(user_messages),
        "assistant_messages": _to_renderable_message_array(assistant_messages),
        "human_ai_pairs": human_ai_pairs,
    }
